caput/fftw.py

```diff
@@ -0,0 +1,356 @@
+"""Fast FFT implementation using FFTW.
+
+This module adds some minor abstraction to use pyfftw in a way
+which seems to be faster than using the `pyfftw.builders` interface,
+and uses the same api as `scipy.fft` and `numpy.fft`.
+
+Only forward and reverse complex->complex transforms
+are currently supported.
+
+Examples
+--------
+The core of this module is the :class:`FFT`, which essentially just
+abstracts the :class:`pyfftw:FFTW` in the simplest way.
+
+>>> import numpy as np
+>>> from caput import fftw
+>>>
+>>> shape = (24, 50)
+>>> x = np.random.rand(*shape) + 1j * np.random.rand(*shape)
+>>>
+>>> fftobj = fftw.FFT(x.shape, x.dtype, axes=-1)
+>>>
+>>> X = fftobj.fft(x)
+>>> xi = fftobj.ifft(X)
+>>>
+>>> np.allclose(x, xi)
+True
+
+The direct API can also be used, although it is slower when doing repeated
+transforms for arrays of the same shape and type because a new :class:`FFT`
+has to be created each time.
+
+References
+----------
+.. https://pyfftw.readthedocs.io
+.. http://www.fftw.org
+
+Classes
+=======
+- :py:class:`FFT`
+
+Functions
+=========
+- :py:meth:`fft`
+- :py:meth:`ifft`
+- :py:meth:`fftconvolve`
+- :py:meth:`fftwindow`
+"""
+
+from __future__ import annotations
+
+import numpy as np
+
+from caput import mpiutil
+
+try:
+    import pyfftw
+except ImportError:
+    raise ImportError(
+        "`pyfftw` is not installed. The `fftw.FFT` class is not available, "
+        "and module functions will use the `scipy` backend. Install `pyfftw` "
+        "via `caput[fftw]`"
+    )
```

Comment:
```suggestion
except ImportError as e:
    raise ImportError(
        "`pyfftw` is not installed. The `fftw.FFT` class is not available, "
        "and module functions will use the `scipy` backend. Install `pyfftw` "
        "via `caput[fftw]`"
    ) from e
```
This message is great.  But I've been trying to become tidier with my exception tracebacks, so here you go:

You should raise the new exception from the old one, to avoid a spurious
```
During handling of the above exception, another exception occurred:
```
 message in the traceback.

(i.e. without `from <...>`, Python assumes the new exception is a completely independent error encountered while handling the original exception, rather than an expected consequence of the prior.)